rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -107,289 +113,144 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
-                        match *e {
-                            RecordSegmentReaderCreationError::UninitializedSegment => {
-                                // This means no compaction has occured, so we can just return whats on the log.
-                                return Ok(MergeMetadataResultsOperatorOutput {
-                                    ids,
-                                    metadata,
-                                    documents,
-                                });
-                            }
-                            RecordSegmentReaderCreationError::BlockfileOpenError(_) => {
-                                error!("Error creating Record Segment: {:?}", e);
-                                return Err(
-                                    MergeMetadataResultsOperatorError::RecordSegmentCreationError(
-                                        *e,
-                                    ),
-                                );
-                            }
-                            RecordSegmentReaderCreationError::InvalidNumberOfFiles => {
-                                error!("Error creating Record Segment: {:?}", e);
-                                return Err(
-                                    MergeMetadataResultsOperatorError::RecordSegmentCreationError(
-                                        *e,
-                                    ),
-                                );
-                            }
-                        }
+                .then_some(log.offset_id)
+            }));
+
+        // Merge the offset ids, assuming the user_offset_ids and filtered_offset_ids are ordered.
+        // This should be a sorted list of unique offset ids.
+        // We are treating the sorted vecs as sets.
+        let merged_oids_holder: Vec<u32>;
+        let merged_offset_ids = match (&input.filtered_offset_ids, &input.user_offset_ids) {
+            (Some(fids), Some(uids)) => {
+                // We take the set intersect of offset ids of those passing the filter and those correspond to user specified ids.
+                merged_oids_holder = merge_sorted_vecs_conjunction(fids, uids);
+                &merged_oids_holder
+            }
+            (Some(oids), None) | (None, Some(oids)) => oids,
+            _ => {
+                // If both filter and user offset ids are None, it suggests user did not specify any filter. We should return all available documents.
+                // Offset ids from the materialized records should be unique, so we only need to sort them.
+                let mut live_log_offset_ids = mat_records
+                    .iter()
+                    .map(|(log, _)| log.offset_id)
+                    .collect::<Vec<_>>();
+                live_log_offset_ids.sort();
+                merged_oids_holder = match &record_segment_reader {
+                    Some(reader) => {
+                        // TODO: We may not need to have the entire offset ids, given limit and offset.
```

Comment:
Why would we get all the offset ids? Assuming 128byte limits and 1M collection, we could have just fetched 128 MB of data? I think we should make this change.